p2p/transport/pipe/pipetransport.go

```diff
@@ -0,0 +1,93 @@
+package pipetransport
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"sync"
+
+	peer "github.com/libp2p/go-libp2p-peer"
+	tpt "github.com/libp2p/go-libp2p-transport"
+	tptu "github.com/libp2p/go-libp2p-transport-upgrader"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+type PipeTransport struct {
+	mlistenchans *sync.RWMutex
+	listenchans  map[string]chan *PipeConn
+
+	upgrader *tptu.Upgrader
+}
+
+var _ tpt.Transport = (*PipeTransport)(nil)
+
+func New(u *tptu.Upgrader) *PipeTransport {
+	return &PipeTransport{
+		mlistenchans: new(sync.RWMutex),
+		listenchans:  make(map[string]chan *PipeConn),
+		upgrader:     u,
+	}
+}
+
+func (t *PipeTransport) closeListener(addr string) {
+	t.mlistenchans.Lock()
+	defer t.mlistenchans.Unlock()
+
+	ch, ok := t.listenchans[addr]
+	if !ok {
+		return
+	}
+	close(ch)
+	delete(t.listenchans, addr)
+}
+
+func (t *PipeTransport) CanDial(addr ma.Multiaddr) bool {
+	protocols := addr.Protocols()
+	return len(protocols) == 1 && protocols[0].Code == ma.P_P2P
+}
+
+func (t *PipeTransport) Protocols() []int {
+	return []int{
+		ma.P_P2P,
+	}
+}
+
+func (t *PipeTransport) Proxy() bool {
+	return false
+}
+
+func (t *PipeTransport) Dial(ctx context.Context, raddr ma.Multiaddr, p peer.ID) (tpt.Conn, error) {
+	t.mlistenchans.RLock()
+	defer t.mlistenchans.RUnlock()
+	raddrStr := raddr.String()
+
+	ch, ok := t.listenchans[raddrStr]
+	if !ok {
+		return nil, fmt.Errorf("no memorylistener for %s", raddrStr)
+	}
+
+	connA, connB := net.Pipe()
+	manetConnA := WrapNetConn(connA, raddr)
+	manetConnB := WrapNetConn(connB, raddr)
+	ch <- manetConnA
+	return t.upgrader.UpgradeOutbound(ctx, t, manetConnB, p)
```

Comment:
We probably want to disable encryption for this, it's a self-pipe after all and it only adds overhead.

Reply:
the upgrader provided can be the "insecure" upgrader which just adds muxing

Reply:
How will this work in practice? Let's take care to avoid the encryption overhead.